File: opencv_learn/line_detection.py
"""
直线检测
"""

import cv2
from cv2.typing import MatLike

from Ahri.Vision.utils.cv2_utils import GREEN, img_show


@img_show("FLD")
def FLD(img: MatLike):
    fld = cv2.ximgproc.createFastLineDetector()
    img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    lines = fld.detect(img_gray)
    plot_img = fld.drawSegments(img, lines, linecolor=GREEN)
    return plot_img


@img_show("LSD")
def LSD(img: MatLike):
    lsd = cv2.createLineSegmentDetector()
    img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    lines, width, prec, nfs = lsd.detect(img_gray)
    img = lsd.drawSegments(img, lines)
    return img


@img_show("EDlines")
def EDlines(img: MatLike):
    ed = cv2.ximgproc.createEdgeDrawing()

    # EDParams
    edparams = cv2.ximgproc.EdgeDrawing.Params()
    edparams.MinPathLength = 50
    edparams.PFmode = False
    edparams.MinLineLength = 10
    edparams.NFAValidation = True
    ed.setParams(edparams)

    img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    ed.detectEdges(img_gray)
    lines = ed.detectLines()
    lines = lines.astype(int)
    for [[x1, y1, x2, y2]] in lines:
        cv2.line(img, (x1, y1), (x2, y2), GREEN, 2, cv2.LINE_AA)
    return img


def main():
    img = cv2.imread(r"D:\Andromeda\Sokyoei\Vision\images\wx_20241120152605.jpg")
    FLD(img)
    LSD(img)
    EDlines(img)


if __name__ == "__main__":
    main()

File: opencv_learn/line_detection.cpp
#include <filesystem>
#include <vector>

#include <opencv2/opencv.hpp>
#include <opencv2/ximgproc.hpp>

#include "Asuka.hpp"
#include "Ahri/Asuka/opencv_utils.hpp"

namespace Ahri {
void FLD(cv::Mat& img) {
    cv::Mat img_gray{img.size(), CV_8UC1};
    cv::Mat plot_img = img.clone();
    std::vector<cv::Vec4f> lines;

    auto fld = cv::ximgproc::createFastLineDetector();
    cv::cvtColor(img, img_gray, cv::COLOR_BGR2GRAY);
    fld->detect(img_gray, lines);
    fld->drawSegments(plot_img, lines);
    // gray ?
    IMG_SHOW(plot_img, "FLD", cv::WINDOW_FREERATIO)
}

void LSD(cv::Mat& img) {
    cv::Mat img_gray{img.size(), CV_8UC1};
    cv::Mat plot_img = img.clone();
    std::vector<cv::Vec4f> lines;

    auto lsd = cv::createLineSegmentDetector();
    cv::cvtColor(img, img_gray, cv::COLOR_BGR2GRAY);
    lsd->detect(img_gray, lines);
    lsd->drawSegments(plot_img, lines);
    IMG_SHOW(plot_img, "LSD", cv::WINDOW_FREERATIO)
}

void EDlines(cv::Mat& img) {
    cv::Mat img_gray{img.size(), CV_8UC1};
    cv::Mat plot_img = img.clone();
    std::vector<cv::Vec4f> lines;

    auto ed = cv::ximgproc::createEdgeDrawing();

    // EDParams
    ed->params.MinPathLength = 50;
    ed->params.PFmode = false;
    ed->params.MinLineLength = 10;
    ed->params.NFAValidation = true;

    cv::cvtColor(img, img_gray, cv::COLOR_BGR2GRAY);
    ed->detectEdges(img_gray);
    ed->detectLines(lines);
    for (auto&& points : lines) {
        cv::line(plot_img, cv::Point(points[0], points[1]), cv::Point(points[2], points[3]), cv::Scalar(0, 255, 0));
    }
    IMG_SHOW(plot_img, "EDlines", cv::WINDOW_FREERATIO)
}
}  // namespace Ahri

int main(int argc, char const* argv[]) {
    auto img_path = std::filesystem::path(SOKYOEI_DATA_DIR) / "Ahri/Popstar Ahri.jpg";
    auto img = cv::imread(img_path.string(), cv::IMREAD_COLOR);

    Ahri::FLD(img);
    Ahri::LSD(img);
    Ahri::EDlines(img);

    return 0;
}
